galaxy/worker/tasks/collection.py

```diff
@@ -0,0 +1,143 @@
+        artifact.delete()
+        task_logger.error('Import Task "{task_id}" failed: {message}'
+                          .format(task_id=task.id, message=str(e)))
+        raise
+
+    warnings, errors = task.get_message_stats()
+    msg = ('Import completed with {warnings} warnings and {errors} errors'
+           .format(warnings=warnings, errors=errors))
+    task_logger.info(msg)
+
+
+def _get_task_logger(task):
+    logger = logging.getLogger('galaxy.worker.tasks.import_collection')
+    return logutils.ImportTaskAdapter(logger, task=task)
+
+
+def _process_collection(artifact, filename, task_logger):
+    with tempfile.TemporaryDirectory() as extract_dir:
+        with artifact.file.open() as pkg_file, \
+                tarfile.open(fileobj=pkg_file) as pkg_tar:
+            pkg_tar.extractall(extract_dir)
+
+        try:
+            collection_info = importer.import_collection(
+                extract_dir, filename, task_logger)
+        except i_exc.ImporterError as e:
+            raise exc.ImportFailed(str(e))
+
+    _log_collection_info(collection_info)
+    return collection_info
+
+
+@transaction.atomic
+def _publish_collection(task, artifact, repository, collection_info):
+    metadata = collection_info.collection_info
+    collection, _ = models.Collection.objects.get_or_create(
+        namespace=task.namespace, name=metadata.name)
+
+    try:
+        version = collection.versions.create(
+            version=metadata.version,
+            metadata=metadata.get_json(),
+            quality_score=collection_info.quality_score,
+        )
+    except IntegrityError:
+        raise exc.VersionConflict('Collection version already exists.')
+
+    relative_path = '{0}-{1}-{2}.tar.gz'.format(
+        metadata.namespace,
+        metadata.name,
+        metadata.version
+    )
+    pulp_models.ContentArtifact.objects.create(
+        artifact=artifact,
+        content=version,
+        relative_path=relative_path,
+    )
+
+    with pulp_models.RepositoryVersion.create(repository) as new_version:
+        new_version.add_content(
+            pulp_models.Content.objects.filter(pk=version.pk)
+        )
+
+    publication = pulp_models.Publication.objects.create(
+        repository_version=new_version,
+        complete=True,
+        pass_through=True,
+    )
+    pulp_models.Distribution.objects.update_or_create(
+        name='galaxy',
+        base_path='galaxy',
+        defaults={'publication': publication},
+    )
+
+    task.imported_version = version
+    task.save()
+
+
+def _log_collection_info(collection_info):
+    log.debug('Collection loaded - metadata={}, quality_score={}'.format(
+        collection_info.collection_info.__dict__,
+        collection_info.quality_score
+    ))
+    for content in collection_info.contents:
+        log.debug('Content: type={} name={} scores={}'.format(
+            content.content_type,
+            content.name,
+            content.scores,
+        ))
```

Comment:
Per the discussion at stand-up this morning with Chris, these new functions are prime candidates to improve the ratio of code-to-tests.

Reply:
These worker collection tests and docstrings are things I should have added in previous PRs, and I can make a note to add after this PR is merged